### PR TITLE
WIP: #5605 fix Chinese search index

### LINKED ISSUE #5605: English words cannot be searched in Chinese word segmentation mode
## Problem:
I can't search for English words in Chinese search mode.

image

## Procedure to reproduce the problem

1. When I set `html_search_language = 'zh'` to use Chinese search, .
2. And add the following string `可以查看 FAQ 模块中 test 部分`.
3. search keyword `test`.

## Error results
When searching keyword `test`, there will be no result display.

## Expected results
There will be test string in search results.

## Sugestions
Why do space characters appear at the end of the word segmentation result?

I look up the source code and send the following regular expression, which will identify the space.
[sphinx/search/zh.py](https://github.com/sphinx-doc/sphinx/blob/16f47cded956bfba55380034ff30988d2ab5ee7c/sphinx/search/zh.py#L236)

Now test as follows：
```
In [3]: re.compile(u'(?u)\\w+[\u0000-\u00ff]').findall("可以查看 FAQ 模块中 test 部分")
Out[3]: ['可以查看 ', 'FAQ ', '模块中 ', 'test ']

```
The result clearly contains the space character, here is the source of the problem.
Can we solve this problem by adjusting regular expressions?

## Environment info

- OS: MacOS 10.14.1 with Python 2.7 in a sphinx specific virtualenv
- Python version: Python 2.7.10
- Sphinx version: 1.8.1
- Browser Chrome

### PR DESCRIPTION
**Pull request by TimKam**
Subject: generate search index for Latin words correctly if search language is Chinese

### Feature or Bugfix
- Bugfix

### Purpose
- As described at #5605
- Conf: ``html_search_language = 'zh'``

### Detail
- Removes trailing (and preceding) spaces from Latin search index terms
- WIP, because a Chinese-speaking person needs to confirm the correctness

### Relates
- Fixes #5605

### CODE REVIEW

**Comment by TimKam** [comment]
@DaGeger can you test once again? What do you think, @tk0miya?

File: sphinx/search/zh.py [hunk 1]
```diff
@@ -250,13 +251,24 @@ def split(self, input):
...
         # type: (unicode) -> bool
         return len(stemmed_word) > 1
 
     def stem(self, word):
         # type: (unicode) -> unicode
+
+        # Don't stem Latin words that are long enough to be relevant for search
+        # if not stemmed, but would be too short after being stemmed
+        # avoids some issues with acronyms
+        should_not_be_stemmed =\
```
Comment (tk0miya) [inline sphinx/search/zh.py]: I feel folding with `\` is a bit ugly. How about using parenthesis like this?
```
should_not_be_stemmed = (
            word in self.latin1 and
            len(word) >= 3 and
            len(self.stemmer.stem(word.lower())) < 3
)
```

**Review by tk0miya — APPROVED** [review]
LGTM with nits

**Review by tk0miya — CHANGES_REQUESTED** [review]
I noticed this should be merged into 1.8 branch because this is a bug fix.

**Review by tk0miya — APPROVED** [review]
LGTM

**Comment by tk0miya** [comment]
Is this still WIP?

**Comment by TimKam** [comment]
Sorry, yes. I wrote some tests locally and found out it needs more work. Will get back to it at the weekend.

**Comment by tk0miya** [comment]
Okay. Please let me know if you finished. I'll review this again :-)

**Comment by TimKam** [comment]
@tk0miya Could you have another look? I needed to add a list that contains all Latin terms and change the regexp that splits Chinese from Latin characters. @DaGeger comments from you regarding the initial problem are also welcome.

**Review by tk0miya — APPROVED** [review]
LGTM!

**Pull request merged**